src/drivers/dw/dma.c

```diff
@@ -526,8 +530,15 @@ static int dw_dma_release(struct dma *dma, int channel)
 	spin_lock_irq(&dma->lock, flags);
 
 	/* get next lli for proper release */
-	p->chan[channel].lli_current =
-		(struct dw_lli2 *)p->chan[channel].lli_current->llp;
+	chan->lli_current = (struct dw_lli2 *)chan->lli_current->llp;
+
+	/* copy leftover data between current and last lli */
+	start_ptr = DW_DMA_LLI_ADDRESS(chan->lli_current, chan->direction);
+	bytes_left = start_ptr - (int32_t)chan->ptr_data.current_ptr;
```

Comment:
What happens if current ptr has bit31 set here (as we are casting to signed) ? start ptr should always be <= current ptr.

Reply:
@lgirdwood Good catch. I've fixed it by doing manual comparison on uints and also changed start_ptr to next_ptr as it's better shows the intentions here.